### PR TITLE
DATACMNS-1699 - Add Embedded annotation.

### PR DESCRIPTION
**Pull request by christophstrobl**
Add core annotation for embedded object support and consider a potential prefix when using `FieldNamingStrategies`.

The actual store specific implementation is done in the individual modules that make sure the mapping context provides means to detect embedded types and provide this information to the `PersistentPropertyPathFactory` now using` `PersistentEntities` instead of the `TypeInformation` to construct mapped paths.

### CODE REVIEW

File: src/main/java/org/springframework/data/mapping/PersistentProperty.java [hunk 1]
```diff
@@ -276,6 +281,24 @@ default Field getRequiredField() {
...
+		return isEntity() && findAnnotation(Embedded.class) != null;
+	}
+
+	/**
+	 * @return {@literal true} if the property generally allows {@literal null} values;
+	 * @since 2.5
+	 */
+	default boolean isNullable() {
+
+		return (isEmbedded() && findAnnotation(Embedded.class).onEmpty().equals(OnEmpty.USE_NULL))
```
Comment (schauder) [inline src/main/java/org/springframework/data/mapping/PersistentProperty.java:298]: When a property is not embedded, it is never nullable. If that is intended, it should be described in the documentation and the superfluous parens around the first `&&` term should be removed. 

Or the expression should get fixed. Also we should have tests for the not embedded case.

**Review by schauder — CHANGES_REQUESTED** [review]

**Comment by dcominottim** [comment]
It'd be really nice if we could further customize this annotation with more parameters or pair it with another annotation to achieve the same effect from `@Embedded` +` @AttributeOverride` from `javax.persistence`. This way, the name of the inline fields could be customized and the embeddable objects could be re-used in different parent objects that need different field names for inline fields.

**Comment by mp911de** [comment]
Embedded objects are conceptually stable regarding their property to field/column/… names. `@Embedded` allows you to provide a prefix to customize physical naming. An example use-case:

```java
class Address {
  String street, city;
}

class Order {

  @Embedded(prefix = "shipping_")
  Address shipping;

  @Embedded(prefix = "billing_")
  Address billing;
}
```

Rendering fields `shipping_street`, `shipping_city`, `billing_street`, and so on.

@dcominottim care to elaborate how else you'd like to customize field names?

**Comment by dcominottim** [comment]
> Embedded objects are conceptually stable regarding their property to field/column/… names. `@Embedded` allows you to provide a prefix to customize physical naming. An example use-case:
> 
> ```java
> class Address {
>   String street, city;
> }
> 
> class Order {
> 
>   @Embedded(prefix = "shipping_")
>   Address shipping;
> 
>   @Embedded(prefix = "billing_")
>   Address billing;
> }
> ```
> 
> Rendering fields `shipping_street`, `shipping_city`, `billing_street`, and so on.
> 
> @dcominottim care to elaborate how else you'd like to customize field names?

Thanks for the reply, @mp911de. I had overlooked the 'prefix' param. It seems to work fine for embedded objects that aggregate multiple properties (such as your `Address` example). I think it should often work for value objects that encapsulate a single property as well (e.g.: `class CompanyName { String companyName; }`), although it'd still be useful for the parent to be able to fully override that -- in JPA, it'd be

```java
class Company {
  @Embedded
  @AttributeOverride(
    name = "companyName", // name of the target attribute
    column = @Column(name = "CMP_NAME") // name that will be used for persistence
  )
  CompanyName name;
}
```

Think of a scenario in which the company you work for has complex rules for data modeling and the data governance team may request you to use specific column/attribute names for certain tables/documents -- and they don't play well with Java/Spring Data attribute naming conventions. I am in such scenario right now and need to use @AttributeOverride in JPA projects.

On a separate note, does Spring Data fully support underscore in attribute names nowadays (as per your prefix example)? It didn't use to. Sorry if too obvious. https://github.com/spring-projects/spring-data-commons/pull/126

**Comment by sseep** [comment]
Thank you for open source contribution.  The feature still be watched?

**Comment by mp911de** [comment]
Superseded by #2294.

**Pull request closed without merging**